tapestry.php

```diff
@@ -349,83 +352,218 @@ function tapestry_tool_log_event()
         } catch (Exception $e) {
             error_log($e->getMessage());
         } finally {
-            update_option($in_progress_option, $no_value);
+            update_option(KalturaUpload::IN_PROGRESS_OPTION, KalturaUpload::NO_VALUE);
+            update_option(KalturaUpload::STOP_UPLOAD_OPTION, KalturaUpload::NO_VALUE, false);
         }
     }
 }
 
-function perform_upload_to_kaltura() {
-    $upload_folder = wp_upload_dir()['path'];
-    
+/**
+ * Uploads videos to Kaltura.
+ *
+ * Video files are transferred up to UPLOAD_BATCH_SIZE at a time.
+ * At the end of each batch, waits synchronously for all videos in the batch to finish converting (or error),
+ * before uploading the next batch.
+ *
+ * @param array $videos     List of video nodes to upload. These should be objects with the following interface:
+ *                          (
+ *                              [tapestryID] => 123,
+ *                              [nodeID] => 123,
+ *                          )
+ *                          Nodes are checked to be videos and to be local Wordpress uploads before being uploaded to Kaltura.
+ * @param bool $use_kaltura_player   Whether to switch uploaded videos to use the Kaltura media player.
+ *
+ * @return int The number of videos that were successfully uploaded.
+ */
+function perform_batched_upload_to_kaltura($videos, $use_kaltura_player)
+{
     $current_date = date('Y/m/d');
 
-    $tapestries = get_posts(['post_type' => 'tapestry',]);
-    $video_nodes = array();
+    $videos_to_upload = create_upload_log($videos);
+    update_upload_log($videos_to_upload);
 
     $kalturaApi = new KalturaApi();
 
-    foreach ($tapestries as $value) {
-        $tapestry = new Tapestry($value->ID);
+    $num_successfully_uploaded = 0;
+    $batch_start = 0;
 
-        foreach ($tapestry->getNodeIds() as $node_id) {
-            $node = new TapestryNode($value->ID, $node_id);
-            $nodeMeta = $node->getMeta();
+    for ($batch_start; $batch_start < count($videos_to_upload); $batch_start += KalturaUpload::UPLOAD_BATCH_SIZE) {
+        // Retrieve fresh value without caching, since we expect the option value to change underneath
+        $GLOBALS['wp_object_cache']->delete(KalturaUpload::STOP_UPLOAD_OPTION, 'options');
+        $stop_requested = get_option(KalturaUpload::STOP_UPLOAD_OPTION);
+        if ($stop_requested === KalturaUpload::YES_VALUE) {
+            break;
+        }
+
+        $batch = array_slice($videos_to_upload, $batch_start, KalturaUpload::UPLOAD_BATCH_SIZE);
+
+        foreach ($batch as $video) {
+            save_video_upload_status($video, $videos_to_upload, UploadStatus::UPLOADING);
+
+            $kaltura_data = null;
+            try {
+                $kaltura_data = $kalturaApi->uploadVideo($video->file, $current_date);
+            } catch (Exception $e) {
+                $error_start = 'Unable to upload video "'.$video->file->name.'" to Kaltura due to: \n';
+
+                error_log($error_start.$e);
```

Comment:
Using the error object `$e` logs both the message and the stack trace, which I thought would be more informative. In the latest change, I have made the code's intention a bit clearer.

(Also, as an aside, although we should use single quotes, I can only print newlines to the error_log using double quotes. This would be why the new code uses double quotes.)